common/src/main/java/io/netty/util/concurrent/Promise.java

```diff
@@ -57,9 +59,55 @@
      */
     boolean setUncancellable();
 
-    @Override
-    Promise<V> addListener(FutureListener<? super V> listener);
+    /**
+     * Cancel the promise and mark the associated future as cancelled.
+     * <p>
+     * If the cancellation was successful it will fail the future with a {@link CancellationException}.
+     *
+     * @param mayInterruptIfRunning set to {@code true} if any thread that might be working on fulfilling the promise
+     *                             is allowed to be interrupted. Otherwise, set to {@code false}.
+     */
+    boolean cancel(boolean mayInterruptIfRunning);
+
+    /**
+     * Get a {@link Future} representation of this promise.
+     *
+     * When this promise succeeds or fails, the future will be completed in a corresponding way.
+     *
+     * The returned future is bound to this specific promise instance, and repeated calls to this method will return
+     * the same future instance.
+     *
+     * @return A {@link Future} representation of this promise.
+     */
+    Future<V> asFuture();
+
+    /**
+     * Returns {@code true} if this promise has already either failed, or been fulfilled successfully.
+     *
+     * @return {@code true} if this promise has been completed, otherwise {@code false}.
+     */
+    boolean isDone();
 
-    @Override
-    <C> Promise<V> addListener(C context, FutureContextListener<? super C, ? super V> listener);
+    /**
+     * Adds the specified listener to this future.
+     * The specified listener is notified when this future is {@linkplain #isDone() done}.
+     * If this future is already completed, the specified listener is notified immediately.
+     *
+     * @param listener The listener to be called when this future completes.
+     *                 The listener will be passed this future as an argument.
+     * @return this future object.
+     */
+    Future<V> addListener(FutureListener<? super V> listener);
```

Comment:
Why is this on Promise? Seems this just encourages the asking-for-bugs approach where code adds a listener to the promise before passing it further down the pipeline, instead of adding it to the returned future.

I'm fine if we just don't want to make that change in this PR, but if we are changing Promise this substantially, then we should fix that gap as well.

Reply:
Use cases like that showed up a lot, so it seemed convenient. Maybe I'll revisit later.

Reply:
Let's punt on this change for now, since we'll keep having Promise extend Future.